src/components/tl/mlx5/mcast/tl_mlx5_mcast_coll.c

```diff
@@ -6,17 +6,249 @@
+    if (UCC_OK == status) {
+        assert(req->comm->ctx != NULL);
+        ucc_tl_mlx5_mcast_mem_deregister(req->comm->ctx, req->rreg);
+        req->rreg = NULL;
+    }
+
+    return status;
+}
+
+static inline int ucc_tl_mlx5_mcast_prepare_bcast(void* buf, int size, int root, void *mr,
+                                                  ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                  ucc_tl_mlx5_mcast_coll_req_t  *req)
+{
+    ucc_status_t             status;
+    ucc_tl_mlx5_mcast_reg_t *reg;
+
+    req->comm    = comm;
+    req->ptr     = buf;
+    req->length  = size;
+    req->root    = root;
+    req->am_root = (root == comm->rank);
+    req->mr      = comm->pp_mr;
+    req->rreg    = NULL;
+    req->proto   = (req->length < comm->max_eager) ? MCAST_PROTO_EAGER : MCAST_PROTO_ZCOPY;
+
+    if (req->am_root) {
+        if (req->proto != MCAST_PROTO_EAGER) {
+           status = ucc_tl_mlx5_mcast_mem_register(comm->ctx, req->ptr, req->length, &reg);
+           if (UCC_OK != status) {
+                return status;
+           }
+           req->rreg = reg;
+           req->mr   = reg->mr;
+        }
+    }
+
+    status = ucc_tl_mlx5_mcast_prepare_reliable(comm, req, req->root);
+    if (ucc_unlikely(UCC_OK != status)) {
+        return status;
+    }
+
+    req->offset      = 0;
+    req->start_psn   = comm->last_psn;
+    req->num_packets = (req->length + comm->max_per_packet - 1)/comm->max_per_packet;
+
+    if (req->num_packets == 0) {
+        req->num_packets = 1;
+    }
+
+    req->last_pkt_len = req->length - (req->num_packets - 1)*comm->max_per_packet;
+
+    assert(req->last_pkt_len > 0 && req->last_pkt_len <= comm->max_per_packet);
+
+    comm->last_psn     += req->num_packets;
+    req->first_send_psn = req->start_psn;
+    req->to_send        = req->am_root ? req->num_packets : 0;
+    req->to_recv        = req->am_root ? 0 : req->num_packets;
+
+    return UCC_OK;
 }
 
-ucc_status_t mcast_coll_do_bcast(void* buf, size_t size, ucc_rank_t root, void *mr, /* NOLINT */
-                                 ucc_tl_mlx5_mcast_coll_comm_t *comm, /* NOLINT */
-                                 ucc_tl_mlx5_mcast_coll_req_t **task_req_handle /* NOLINT */)
+ucc_status_t ucc_tl_mlx5_mcast_coll_do_bcast(void* buf, int size, int root, void *mr,
+                                             ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                             ucc_tl_mlx5_mcast_coll_req_t **task_req_handle)
 {
-    return UCC_ERR_NOT_SUPPORTED;
+    ucc_status_t                  status;
+    ucc_tl_mlx5_mcast_coll_req_t *req;
+
+    tl_debug(comm->lib, "MCAST bcast start, buf %p, size %d, root %d, comm %d, "
+             "comm_size %d, am_i_root %d comm->psn = %d \n",
+             buf, size, root, comm->comm_id, comm->commsize, comm->rank ==
+             root, comm->psn );
+
+    req = ucc_calloc(1, sizeof(ucc_tl_mlx5_mcast_coll_req_t), "mcast_req");
```

Comment:
are you sure you want to allocate requests every time? Usually we use memory pools for requests

Reply:
I'll add this feature in the future PRs.